src/Compiler/Checking/TailCallChecks.fs

```diff
@@ -0,0 +1,868 @@
+open FSharp.Compiler.Features
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.TcGlobals
+open FSharp.Compiler.TypedTree
+open FSharp.Compiler.TypedTreeBasics
+open FSharp.Compiler.TypedTreeOps
+open FSharp.Compiler.TypeRelations
+
+let PostInferenceChecksStackGuardDepth = GetEnvInteger "FSHARP_TailCallChecks" 50
+
+//--------------------------------------------------------------------------
+// check environment
+//--------------------------------------------------------------------------
+
+type env =
+    {
+        /// Values in module that have been marked [<TailCall>]
+        mustTailCall: Zset<Val>
+
+        /// Recursive expressions of [<TailCall>] attributed values
+        mustTailCallExprs: Map<Stamp, Expr>
+    }
+
+    override _.ToString() = "<env>"
+
+let (|ValUseAtApp|_|) e =
+    match e with
+    | InnerExprPat (Expr.App(funcExpr = InnerExprPat (Expr.Val (valRef = vref; flags = valUseFlags))) | Expr.Val (valRef = vref
+                                                                                                                  flags = valUseFlags)) ->
+        Some(vref, valUseFlags)
+    | _ -> None
+
+type IsTailCall =
+    | Yes of bool // true indicates "has unit return type and must return void"
+    | No
+
+    static member private IsVoidRet (g: TcGlobals) (v: Val) =
+        match v.ValReprInfo with
+        | Some info ->
+            let _tps, tau = destTopForallTy g info v.Type
+
+            let _curriedArgInfos, returnTy =
+                GetTopTauTypeInFSharpForm g info.ArgInfos tau v.Range
+
+            isUnitTy g returnTy
+        | None -> false
+
+    static member YesFromVal (g: TcGlobals) (v: Val) =
+        IsTailCall.Yes(IsTailCall.IsVoidRet g v)
+
+    static member YesFromExpr (g: TcGlobals) (expr: Expr) =
+        match expr with
+        | ValUseAtApp (valRef, _) -> IsTailCall.Yes(IsTailCall.IsVoidRet g valRef.Deref)
+        | _ -> IsTailCall.Yes false
+
+    member x.AtExprLambda =
+        match x with
+        // Inside a lambda that is considered an expression, we must always return "unit" not "void"
+        | IsTailCall.Yes _ -> IsTailCall.Yes false
+        | IsTailCall.No -> IsTailCall.No
+
+let IsValRefIsDllImport g (vref: ValRef) =
+    vref.Attribs |> HasFSharpAttributeOpt g g.attrib_DllImportAttribute
+
+type cenv =
+    {
+        stackGuard: StackGuard
+
+        g: TcGlobals
+
+        amap: Import.ImportMap
+
+        reportErrors: bool
+    }
+
+    override x.ToString() = "<cenv>"
+
+//--------------------------------------------------------------------------
+// approx walk of type
+//--------------------------------------------------------------------------
+
+/// Indicates whether an address-of operation is permitted at a particular location
+[<RequireQualifiedAccess>]
+type PermitByRefExpr =
+    /// Permit a tuple of arguments where elements can be byrefs
+    | YesTupleOfArgs of int
+
+    /// Context allows for byref typed expr.
```

Comment:
Byref specifics are really not my forte - just curious how mutually exclusive are the below?
I mean, Yes, YesReturnable and YesReturnableNonLocal. 

If they are, would it make more sense to have a subtype for Yes instead?

Reply:
Sure, I am okay with that. Maybe you can then put a comment what this aligns with, or some short explanation. :)

Reply:
Done